### PR TITLE
Equal store eliminator.

### PR DESCRIPTION
**Pull request by chriseth**
This removes subsequent stores to the same location with the same value, even if reads happen in between.

### CODE REVIEW

**Comment by chriseth** [comment]
Partly solves https://github.com/ethereum/solidity/issues/12211

File: test/libsolidity/semanticTests/structs/struct_copy.sol
Comment (axic) [inline test/libsolidity/semanticTests/structs/struct_copy.sol:41]: Seems to be working 😅

Comment (chriseth) [reply test/libsolidity/semanticTests/structs/struct_copy.sol:41]: :shrug:

Comment (chriseth) [reply test/libsolidity/semanticTests/structs/struct_copy.sol:41]: The assembly is actually shorter. My guess is that the removed mstore instructions makes it impossible for the opcode-based optimizer to remove the multiple keccak calls...

File: libyul/optimiser/EqualStoreEliminator.cpp
Comment (Marenz) [inline libyul/optimiser/EqualStoreEliminator.cpp]: As far as I can tell we only access the `const` member `dialect` for context. Should the parameter be a const ref maybe?

File: libyul/optimiser/EqualStoreEliminator.cpp
Comment (Marenz) [inline libyul/optimiser/EqualStoreEliminator.cpp:55]: Not related to the PR changes, but the name `isSimpleStore` is very confusing given that it returns two variables. Without reading the docs to that function I'd have no idea what's going on here...

Comment (Marenz) [reply libyul/optimiser/EqualStoreEliminator.cpp:55]: Something like `extractStoreArguments` might be more understandable?

File: libyul/optimiser/EqualStoreEliminator.h
Comment (Marenz) [inline libyul/optimiser/EqualStoreEliminator.h]: Are the functions `tryResolve` and `tryEvaluateKecak` implemented and called anywhere?

**Review by Marenz — CHANGES_REQUESTED** [review]

File: libyul/optimiser/EqualStoreEliminator.h
Comment (cameel) [inline libyul/optimiser/EqualStoreEliminator.h:35]: > without literal arguments

Just a question: which optimizer step replaces literal arguments with variables?

File: libyul/optimiser/EqualStoreEliminator.cpp
Comment (cameel) [inline libyul/optimiser/EqualStoreEliminator.cpp:59]: Isn't this condition a bit too simple?

The docstring says that this is best used when there are no literal arguments. So let's take this code:
```solidity
    mstore(0, 0)
    mstore(0, 0)
```
`ExpressionSplitter` seems to remove literal arguments and it turns the code to this:
```solidity
    let _1 := 0
    let _2 := 0
    mstore(_2, _1)
    let _3 := 0
    let _4 := 0
    mstore(_4, _3)
```
but it this would defeat this optimization.

File: test/libyul/YulOptimizerTestCommon.cpp
Comment (cameel) [inline test/libyul/YulOptimizerTestCommon.cpp]: `ForLoopInitRewriter` is listed in the docstring as a dependency too.

**Review by cameel — COMMENTED** [review]
Documentation for the new step is missing.

Also maybe a few more tests beyond the one from the issue might be useful. We can see the effect in full suite tests but I think this also needs some dedicated ones.

File: libyul/optimiser/EqualStoreEliminator.cpp
Comment (cameel) [reply libyul/optimiser/EqualStoreEliminator.cpp:55]: I think that removing `auto` might help.

Too bad that a destructuring assignment like `auto [location, value] =` won't work here due to the `optional`.

File: libyul/optimiser/EqualStoreEliminator.cpp
Comment (chriseth) [reply libyul/optimiser/EqualStoreEliminator.cpp]: This is a standard interface, but I can check.

File: libyul/optimiser/EqualStoreEliminator.cpp
Comment (chriseth) [reply libyul/optimiser/EqualStoreEliminator.cpp:59]: Yes, you also need to call the CSE to make this work properly, but I'm actually currently wondering whether we check somewhere that the value is an ssa variable with a movable value, this should be another condition.

Comment (chriseth) [reply libyul/optimiser/EqualStoreEliminator.cpp:59]: Ah no, it's fine - we only replace variables by variables and we clear everything whenever there was an assignment to the variable.

File: libyul/optimiser/EqualStoreEliminator.h
Comment (chriseth) [reply libyul/optimiser/EqualStoreEliminator.h:35]: the expression splitter

File: libyul/optimiser/EqualStoreEliminator.h
Comment (chriseth) [reply libyul/optimiser/EqualStoreEliminator.h]: Ah, I think that was a copy-past-leftover.

File: libyul/optimiser/EqualStoreEliminator.cpp
Comment (cameel) [reply libyul/optimiser/EqualStoreEliminator.cpp:59]: Can you mention it in the docstring? A comment in this function would not hurt either. This kind of assumption is not easy to notice other than from the fact that the code looks way too simple for that it does so there must be some "catch" somewhere.

File: libyul/optimiser/EqualStoreEliminator.h
Comment (cameel) [inline libyul/optimiser/EqualStoreEliminator.h:35]: I'd make it more explicit. "Works best" does not really tell me what I can expect if my code does not satisfy this.
```suggestion
 * Does not optimize parts of code with variables not in SSA form or with calls with literal arguments.
```

File: libyul/optimiser/EqualStoreEliminator.cpp
Comment (chriseth) [reply libyul/optimiser/EqualStoreEliminator.cpp:59]: What exactly do you want me to mention? These are all properties of the base class.

File: libyul/optimiser/EqualStoreEliminator.h
Comment (chriseth) [reply libyul/optimiser/EqualStoreEliminator.h:35]: But that is exactly what is written there, isn't it? We use this wording in all the other steps as well.

File: libyul/optimiser/UnusedStoreBase.cpp
Comment (hrkrshnn) [inline libyul/optimiser/UnusedStoreBase.cpp]:
```suggestion
```

**Review by hrkrshnn — COMMENTED** [review]
Looks good. I'll commit a couple of tests that I wrote locally while testing this.

Edit: also updated the failing chromosome test.

File: libyul/optimiser/UnusedStoreBase.cpp
Comment (hrkrshnn) [reply libyul/optimiser/UnusedStoreBase.cpp]: Actually, will update this myself.

File: test/libsolidity/semanticTests/externalContracts/deposit_contract.sol
Comment (hrkrshnn) [inline test/libsolidity/semanticTests/externalContracts/deposit_contract.sol:181]:
```diff
diff -u /tmp/develop.yul /tmp/eq.yul
--- /tmp/develop.yul	2022-01-03 17:32:01.880093589 +0530
+++ /tmp/eq.yul	2022-01-03 17:31:28.243155455 +0530
@@ -546,7 +546,6 @@
                 calldatacopy(pos, value0, value1)
                 let _1 := add(pos, value1)
                 mstore(_1, 0)
-                mstore(_1, 0)
                 end := add(_1, 16)
             }
             function calldata_array_index_range_access_bytes_calldata_3157(offset, length) -> offsetOut, lengthOut
@@ -572,8 +571,6 @@
                 let _1 := add(pos, value1)
                 mstore(_1, /** @src 0:6663:6664  "0" */ 0x00)
                 /// @src 0:4559:9399  "contract DepositContract is IDepositContract, ERC165 {..."
-                mstore(_1, /** @src 0:6663:6664  "0" */ 0x00)
-                /// @src 0:4559:9399  "contract DepositContract is IDepositContract, ERC165 {..."
                 end := add(_1, 32)
             }
             function abi_encode_packed_bytes32_bytes_calldata(pos, value0, value1, value2) -> end
@@ -937,7 +934,7 @@
                 mstore8(memory_array_index_access_bytes_3175(memPtr), /** @src 0:9377:9390  "bytesValue[0]" */ byte(/** @src -1:-1:-1 */ 0, /** @src 0:9377:9390  "bytesValue[0]" */ _1))
             }
         }
-        data ".metadata" hex"a364697066735822122006c8a10f3f04d8c69b52af5dbc3a82224acdfaa0224b0f3ed0cf38c99a4f0c7e6c6578706572696d656e74616cf564736f6c637823302e382e31322d63692e323032312e362e32332b636f6d6d69742e36383439373734620062"
+        data ".metadata" hex"a3646970667358221220a5a5d6e1a1822c1fcf45d9e6ac103252bac21d4af4c04b1a690185ad3147faf46c6578706572696d656e74616cf564736f6c63782b302e382e31322d646576656c6f702e323032322e312e332b636f6d6d69742e33633634333238612e6d6f64006a"
     }
 }
 

Diff finished.  Mon Jan  3 17:32:12 2022
```

Here's the diff!

File: test/libsolidity/semanticTests/structs/struct_copy.sol
Comment (chriseth) [reply test/libsolidity/semanticTests/structs/struct_copy.sol:41]:
```diff
--- /tmp/ir_old.yul	2022-01-03 15:46:30.242337855 +0100
+++ /tmp/ir_opt.yul	2022-01-03 15:41:35.197221133 +0100
@@ -79,34 +79,26 @@
                 mstore(_1, var_k)
                 mstore(0x20, _1)
                 sstore(keccak256(_1, 0x40), 0x01)
-                mstore(_1, var_k)
-                mstore(0x20, _1)
                 sstore(add(keccak256(_1, 0x40), 0x01), 0x03)
-                mstore(_1, var_k)
-                mstore(0x20, _1)
                 sstore(add(keccak256(_1, 0x40), 2), 0x04)
-                mstore(_1, var_k)
-                mstore(0x20, _1)
                 sstore(add(keccak256(_1, 0x40), 0x03), 2)
                 var := 0x01
             }
             function fun_copy(var_from, var_to) -> var_
             {
-                let _1 := 0x00
-                mstore(_1, var_from)
-                mstore(0x20, _1)
-                let dataSlot := keccak256(_1, 0x40)
-                mstore(_1, var_to)
-                mstore(0x20, _1)
-                let dataSlot_1 := keccak256(_1, 0x40)
+                mstore(0x00, var_from)
+                mstore(0x20, 0x00)
+                let dataSlot := keccak256(0x00, 0x40)
+                mstore(0x00, var_to)
+                let dataSlot_1 := keccak256(0x00, 0x40)
                 if iszero(eq(dataSlot_1, dataSlot))
                 {
                     sstore(dataSlot_1, sload(dataSlot))
                     let memberSlot := add(dataSlot_1, 1)
-                    let _2 := add(dataSlot, 1)
-                    if iszero(eq(memberSlot, _2))
+                    let _1 := add(dataSlot, 1)
+                    if iszero(eq(memberSlot, _1))
                     {
-                        sstore(memberSlot, sload(_2))
+                        sstore(memberSlot, sload(_1))
                         sstore(add(dataSlot_1, 2), sload(add(dataSlot, 2)))
                     }
                     sstore(add(dataSlot_1, 3), sload(add(dataSlot, 3)))
@@ -119,18 +111,12 @@
                 mstore(_1, var_k)
                 mstore(0x20, _1)
                 var_a := sload(keccak256(_1, 0x40))
-                mstore(_1, var_k)
-                mstore(0x20, _1)
                 var_x := sload(add(keccak256(_1, 0x40), 1))
-                mstore(_1, var_k)
-                mstore(0x20, _1)
                 var_y := sload(add(keccak256(_1, 0x40), 2))
-                mstore(_1, var_k)
-                mstore(0x20, _1)
                 var_c := sload(add(keccak256(_1, 0x40), 3))
             }
         }
-        data ".metadata" hex"a36469706673582212201575720a9612338487e4ea44beef0f99559f8ddb767d3b3a8f88b30e8203f9ee6c6578706572696d656e74616cf564736f6c634300080c0041"
+        data ".metadata" hex"a36469706673582212201300a8828ade0dfe77d7dbd65f00b07037165a285fa1ba1a0c986a59b74ec9b66c6578706572696d656e74616cf564736f6c634300080c0041"
     }
 }
```

Comment (chriseth) [reply test/libsolidity/semanticTests/structs/struct_copy.sol:41]: The main change is that zero constants are now inlined, so I would say it's not an issue.

File: libyul/optimiser/EqualStoreEliminator.cpp
Comment (hrkrshnn) [inline libyul/optimiser/EqualStoreEliminator.cpp]:
```suggestion
```

File: libyul/optimiser/EqualStoreEliminator.cpp
Comment (hrkrshnn) [inline libyul/optimiser/EqualStoreEliminator.cpp]:
```suggestion
```

File: libyul/optimiser/EqualStoreEliminator.cpp
Comment (hrkrshnn) [inline libyul/optimiser/EqualStoreEliminator.cpp]:
```suggestion
```

File: libyul/optimiser/EqualStoreEliminator.cpp
Comment (hrkrshnn) [inline libyul/optimiser/EqualStoreEliminator.cpp]:
```suggestion
```

**Review by hrkrshnn — COMMENTED** [review]
Going to fix it myself.

**Review by hrkrshnn — DISMISSED** [review]
LGTM! Added a couple more tests and removed some more unused headers in the last commit. Feel free to squash the commits.

Would be nice to get a yes from @bshastry before merging, with regard to changes in fuzzer.

File: test/libsolidity/semanticTests/structs/struct_copy.sol
Comment (hrkrshnn) [reply test/libsolidity/semanticTests/structs/struct_copy.sol:41]: Should perhaps implement evaluating `keccak` for multi words (at least 2 words).

**Review by bshastry — COMMENTED** [review]
LGTM wrt fuzzer once https://github.com/ethereum/solidity/pull/12276 is merged

**Pull request merged**
